src/lapack/gpu/lacpy.cu

```diff
@@ -105,6 +105,80 @@ __global__ void lacpy(cublasFillMode_t uplo, const unsigned m, const unsigned n,
 }
 }
 
+static whip::memcpy_kind get_lacpy_memcpy_kind(const void* src, const void* dst) {
+// If HIP is version 5.6 or newer, avoid the use of hipMemcpyDefault as it is
+// buggy with 2D memcpy.  Instead try to infer the memory type using
+// hipPointerGetAttributes. See:
+// - https://github.com/ROCm/clr/commit/56daa6c4891b43ec233e9c63f755e3f7b45842b4
+// - https://github.com/ROCm/clr/commit/d3bfb55d7a934355257a72fab538a0a634b43cad
+//
+// Note that hipMemoryTypeManaged is not available in older versions, but it is
+// already available in 5.3 so we don't do a separate check for availability.
+#if defined(DLAF_WITH_HIP) && HIP_VERSION >= 50600000
+  constexpr auto get_memory_type = [](const void* p) {
+    hipPointerAttribute_t attributes{};
+    hipError_t status = hipPointerGetAttributes(&attributes, p);
+    if (status == hipErrorInvalidValue) {
+      // If HIP returns hipErrorInvalidValue we assume it's due
+      // to HIP not recognizing a non-HIP allocated pointer as
+      // host memory, and we assume the type is host.
+      return hipMemoryTypeHost;
+    }
+    else if (status != hipSuccess) {
+      throw whip::exception(status);
+    }
+
+    switch (attributes.type) {
+#if HIP_VERSION >= 60000000
+      case hipMemoryTypeUnregistered:
+        [[fallthrough]];
+#endif
+      case hipMemoryTypeHost:
+        return hipMemoryTypeHost;
+      case hipMemoryTypeArray:
+        [[fallthrough]];
+      case hipMemoryTypeDevice:
+        return hipMemoryTypeDevice;
+      case hipMemoryTypeUnified:
+        return hipMemoryTypeUnified;
+      case hipMemoryTypeManaged:
+        return hipMemoryTypeManaged;
+      default:
+        DLAF_UNREACHABLE_PLAIN;
+    }
+  };
+
+  hipMemoryType src_type = get_memory_type(src);
+  hipMemoryType dst_type = get_memory_type(dst);
+
+  if (src_type == hipMemoryTypeDevice && dst_type == hipMemoryTypeHost) {
+    return whip::memcpy_device_to_host;
+  }
+  else if (src_type == hipMemoryTypeHost && dst_type == hipMemoryTypeDevice) {
+    return whip::memcpy_host_to_device;
+  }
+  else if (src_type == hipMemoryTypeDevice && dst_type == hipMemoryTypeDevice) {
+    return whip::memcpy_device_to_device;
+  }
+  else if (src_type == hipMemoryTypeManaged || src_type == hipMemoryTypeUnified ||
+           dst_type == hipMemoryTypeManaged || dst_type == hipMemoryTypeUnified) {
+    return whip::memcpy_default;
+  }
+  else if (src_type == hipMemoryTypeHost && dst_type == hipMemoryTypeHost) {
+    DLAF_ASSERT(
+        false,
+        "Attempting to do a HIP lacpy with host source and destination, use the CPU lacpy instead");
+  }
```

Comment:
Currently this assertion is only triggered with HIP 5.6 and newer, but theoretically it's useful also with other versions, including with CUDA. I don't know if it's worth it to always do this assertion though. I don't think it's a big deal to not check it (as we do right now on `master`).